Add infra-name annotations to kube gen/play

umohnani8 opened the pull request:
Add io.podman.annotations.infra.name annotation to kube play so
users can set the name of the infra container created.
When a pod is created with --infra-name set, the generated
kube yaml will have an infraName annotation set that will
be used when playing the generated yaml with podman.

Fixes https://github.com/containers/podman/issues/18312
<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Set an infra name annotation in the generated yaml when the `pod create` command has --infra-name set.
This annotation can also be used with `kube play` when wanting to customize the infra container name.
```

Luap99 commented on docs/source/markdown/podman-kube-play.1.md.in:
```diff
@@ -190,6 +190,10 @@ Tear down the volumes linked to the PersistentVolumeClaims as part of --down
 
+#### **--infra-name**
+
+Name for the infra container when creating the pod. If not given, the default name of podID-infra is used.
```
I think one way to set is is best, if we only use the annotation in the yaml it should work correctly for multiple pods in one file and we do not worry about testing multiple combinations.

umohnani8 replied:
Yes it will error on creation, I think we should support this for only the one pod use case and note that in the docs.

Luap99 replied:
But why have two ways to do the same thing? If a user can already set it via yaml annotation then there is no need for a cli option IMO.

vrothberg replied:
The request AFAIUI is about adding the --infra-name flag (https://github.com/containers/podman/issues/18312).

vrothberg commented on libpod/define/annotations.go:
```diff
@@ -145,6 +145,10 @@ const (
 
+	// InfraNameAnnotation is used by generate and play kube when the infra container is set by the user during
+	// pod creation
+	InfraNameAnnotation = "io.podman.annotations.infra.name"
```
> having the annotation would help translate a pod create --infra-name command to the generated kube yaml

I don't think we need the annotation.  By looking at the name of the pod's infra container, we know whether a custom name was used (i.e., when `name != $podID-infra`).

danishprakash replied:
> By looking at the name of the pod's infra container, we know whether a custom name was used (i.e., when name != $podID-infra).

But I guess then you'd not be able to replay the generated kube with the custom name because there seems to be no inbuilt `infraName` attribute. That's where this annotation comes in, even on regenerate, podman can effectively recreate the custom infra name.

umohnani8 commented:
I updated the PR so that we are only using the annotation `io.podman.annotations.infra.name` for kube generate and kube play. This solves the issue where we have multiple pod definitions in one yaml and we want to set a cusotmized infra name for each of them. 
I have added this information to the docs as well. There is no longer a `--infra-name` flag for podman kube play.

vrothberg commented on test/e2e/play_kube_test.go:
```diff
@@ -5879,4 +5879,80 @@ EXPOSE 2004-2005/tcp`, ALPINE)
+		pod.WaitWithDefaultTimeout()
+		Expect(pod).Should(Exit(0))
+
+		ctr := podmanTest.Podman([]string{"run", "-d", "--pod", podName, ALPINE, "top"})
```
Nit: using `create` instead of `run` will make the test slightly faster.

vrothberg replied:
Same below.

umohnani8 replied:
Fixed

vrothberg reviewed (commented):
LGTM

rhatdan commented:
/lgtm
